Convert functions to use code points

MonoidMusician opened the pull request:
This is more accurate because `CodePoint` represents **any** unicode character, whereas `Char` only works with non-astral characters, due to its JavaScript representation. Right?

I'll probably submit a PR to `-strings` to add an equivalent of `charPoint` directly to Data.String.CodePoints, which will clean this up a bit.

MonoidMusician committed:
Convert functions to use code points

This is more accurate because`CodePoint` represents any unicode character, whereas `Char` only works with non-astral characters, due to its JavaScript representation.

cdepillabout commented:
@MonoidMusician Thanks for this PR!

To be honest, I'm not super knowledgable about unicode.  Would you be able to link me to something explaining the difference between Javascript `Char`s and actual unicode characters?

Before I merge this, could you do the following things:

- update the tests with unit tests that are not possible to write when using `Char`, but are now possible when using `CodePoint`
- update the examples in the docoments for each function (for instance the `isPunctuation` function: https://github.com/purescript-contrib/purescript-unicode/pull/15/files#diff-a8371e327cfff9b134c465f53ecdd381L407)
- create a `Changelog.md`, add it to `package.json` (or `bower.json` or whereever it needs to go), and add an explanation about changing `Char` to `CodePoint`, linking back to this PR

Also, [purescript-parsing](https://github.com/purescript-contrib/purescript-parsing) depends on purescript-unicode.  After this is merged, would you be willing to send a small PR to purescript-parsing updating their uses of purescript-unicode?

cdepillabout commented:
@michaelficarra, would it be possible for you to take a quick glance at this PR and make sure everything looks sane?

It looks like you're the author of `Data.String.CodePoint`, so I was thinking you might be the most qualified to tell whether this use of it is good.

In case you haven't see `purescript-unicode` before, it is basically a straight copy of Haskell's [`Data.Char`](https://hackage.haskell.org/package/base-4.10.0.0/docs/Data-Char.html)  module.

cdepillabout commented:
Also, I wonder if it makes sense to change the module name from `Data.Char.Unicode`, to something like `Data.CodePoint.Unicode`.

michaelficarra commented on src/Data/Char/Unicode.purs:
```diff
@@ -478,41 +487,43 @@ isSymbol c =
-toUpper :: Char -> Char
-toUpper = fromCharCode <<< uTowupper <<< toCharCode
+toUpper :: CodePoint -> CodePoint
+toUpper = modify uTowupper
```
`toUpper` can't be `CodePoint -> CodePoint`, it must be `CodePoint -> String`. But maybe that's something we should address separately, since I see it's currently broken anyway.

MonoidMusician replied:
Okay ... Is this because a some precomposed characters don't have direct upper case equivalents and would need to be decomposed?

michaelficarra replied:
Possibly, but also see https://unicode.org/faq/casemap_charprop.html#11

MonoidMusician replied:
Okay, cool ... I'm not familiar with the representation of the rules, so I wouldn't know how to fix that. Also, would we want to return `String` or `Array CodePoint`? (maybe the latter for the the `Internal` module?)

michaelficarra replied:
`Array CodePoint` sounds good.

MonoidMusician replied:
currently it looks like `toUpper` does not change the eszett.

cdepillabout replied:
@MonoidMusician could you also add some documentation to `toUpper` explaining why the type is `CodePoint -> Array CodePoint`?

Also, I'm wondering if it would be helpful to have a function like `unsafeToUpper :: CodePoint -> CodePoint` for people who want to easily map over a list of codepoints, without worrying about being 100% correct.  Its use should be discouraged of course.

michaelficarra commented:
I don't like the use of `unsafePartial`, but yes, this looks like the right direction. Also, I agree that something like `charPoint` should be built in.

MonoidMusician commented:
@cdepillabout Thanks for the suggestions!

According the the documentation for Prim.Char [1], it represents one UTF-16 code unit, i.e. all _non-astral_ code points. But CodePoint is a newtype around Int so it can represent any Unicode value.

``Int `superset` CodePoint `superset` Char`` I believe

I should make a PR for `-strings` to make most of those changes ... maybe `codePoint{To,From}Char` would be a good name instead of `charPoint`?

[1] https://pursuit.purescript.org/builtins/docs/Prim#t:Char

cdepillabout commented:
> Also, I wonder if it makes sense to change the module name from `Data.Char.Unicode`, to something like `Data.CodePoint.Unicode`.

After thinking about it some more, I think this might be a good change to make.

It would make it possible for us to go back and re-add a `Data.Char.Unicode` module that works on `Char` instead of `CodePoint`.  That would make it easy for people to use without having to figure out the relationship between `Char` and `CodePoint`.  (I'm assuming that most users

The documentation for `Data.Char.Unicode` should point out its deficiencies, and also point users to the more-correct `Data.CodePoint.Unicode` module.

michaelficarra commented:
Why even have the `Data.Char` one? We have code unit based functions in `-strings` only for legacy and are considering moving around the modules to make `CodePoint` the default.

cdepillabout commented:
@michaelficarra 

> We have code unit based functions in -strings only for legacy and are considering moving around the modules to make CodePoint the default.

Does "code unit" mean PureScript's `Char`?

I was thinking that there would be a lot of [beginner?] PureScript programmers that are using `Char` (since it is explained in the PureScript book and available in `Prim`).  Many of them may want functions like `toUpper`, `toLower`, `isSymbol` etc they can easily use, without having to figure out the `CodePoint` type.  That's why we should also provide a `Data.Char.Unicode` module in addition to (my proposed) `Data.CodePoint.Unicode`.

However, there will certainly be people that want to handle unicode correctly.  They can use the `Data.CodePoint.Unicode` module.  The `Data.Char.Unicode` should specifically recommend using `Data.CodePoint.Unicode`.

-------------------------------------------

However, if the rest of the PureScript community is moving to completely use `CodePoint` instead of `Char`, then I think that you are correct.  We don't really need the `Data.Char.Unicode` module.

MonoidMusician committed:
Change namespace to Data.CodePoint.Unicode

MonoidMusician committed:
Update docs too

MonoidMusician commented:
I will update this PR once this is merged and released: https://github.com/purescript/purescript-strings/pull/92

It looks like the mapping of “ß” `00DF` to “SS” `0073 0073` is not specified in `UnicodeData.txt` but rather [`CaseFolding.txt`](http://www.unicode.org/Public/10.0.0/ucd/CaseFolding.txt). Should we parse that as well and use that for case data? Like you said, `CodePoint -> Array CodePoint` makes sense here, and we can also expose `String -> String` by using `concatMap` with the appropriate conversions.

```
# UnicodeData.txt specifies:
00DF;LATIN SMALL LETTER SHARP S;Ll;0;L;;;;;N;;;;;
# CaseFolding.txt specifies:
00DF; F; 0073 0073; # LATIN SMALL LETTER SHARP S
```

I've started my own standard parsing library so if I could probably attempt to convert our parser here to PS from awk: https://github.com/MonoidMusician/purescript-uievents-key/blob/master/test/Generate.purs

Also should we generate it from the latest 10.0.0 instead of 6.0.0 linked in the README? I'm not sure how significant the changes are (more emoji anyone?? 😄)

And my last point: I renamed it to `Data.CodePoints.Unicode` but the strings package uses `Data.String.CodePoints` ... should I go to `Data.String.CodePoints.Unicode` to be consistent with that? I personally don't think that keeping `Char` around makes much sense, especially since we will be able to easily upcast that once the PR is merged.

Thanks for the feedback from both of you!

michaelficarra commented:
> Should we parse that as well and use that for case data?

Yes, if you want to provide case mapping functions, you should base them on that table.

> Also should we generate it from the latest 10.0.0 instead of 6.0.0 linked in the README?

Yes, but stay consistent within major releases of purescript-unicode.

> should I go to `Data.String.CodePoints.Unicode` to be consistent with that?

No, see https://github.com/purescript/purescript-strings/issues/95.

cdepillabout commented:
It looks like there has been some discussion about moving to CodePoints on https://github.com/purescript/purescript-strings/issues/95, so it seems like it might be a good time to revisit this issue!

kritzcreek commented:
I'm going to need a response to this "quickly" or I'll make a 0.12 release with the old API and we'll need to make another breaking release to incorporate this PR.

cdepillabout commented:
@MonoidMusician Does this PR need to be updated?  Or can it be merged in as-is?

MonoidMusician commented:
I am starting to work on updating it :) Special casing might be a separate PR though ... maybe we should keep `toUpper`/`toLower` as-is (or rename now), and add new functions for [Special Casing](https://unicode.org/Public/UNIDATA/SpecialCasing.txt)?

kritzcreek commented:
@MonoidMusician For anything that you do, make sure you do it against the compiler/0.12 branch. This package-set should have all the dependencies you need: https://github.com/kRITZCREEK/package-sets/tree/test-core

EDIT:

fromCharCode returns a Maybe now, but since this library is pretty low-level you might want to redefine a non-Maybe version with the FFI and use that whenever you're sure you'll stay within bounds.

MonoidMusician commented:
Hm, the dev dependency `purescript-spec` isn't updated ...

MonoidMusician committed:
Update for 0.12

MonoidMusician committed:
Update tests

They work with local modifications ...

MonoidMusician committed:
Clean up docs again, fix out of bounds error

MonoidMusician committed:
Update to Unicode 10.0.0

MonoidMusician committed:
Fix import warnings

MonoidMusician commented:
I had to fix up the testing libraries locally, but the tests pass here now. @kRITZCREEK suggested using `purescript-assert` so we don't have to "wait" on spec being updated ...

MonoidMusician committed:
YOLO

jk tests still pass

hdgarrood commented:
I'm interested in helping move this forward; does anyone have suggestions for how I could do that?

thomashoneyman commented:
@MonoidMusician We're making breaking changes for libraries for the upcoming 0.14 release. I know it's been a long, long time since you've looked at this, but I'm curious if you have any advice or suggestions for folks who would like to move this along and get it in for the upcoming breaking release. Thanks!

MonoidMusician committed:
Merge remote-tracking branch 'purescript-contrib/main'

MonoidMusician committed:
Add full case conversion algorithms

These variants may map one code point to more than one code point.

MonoidMusician commented:
I got around to this again! I added a script to scrape data from both CaseFolding.txt and SpecialCasing.txt – it turns out that the former only specifies canonical casings (and has full and simple variants), while the latter is the full lower/title/upper mappings that we were looking for.

Are we happy with the types and naming?

I need to think more about what `toTitle :: String -> String` will look like …

Should also update to a later version of the standard.

MonoidMusician committed:
Update to Unicode 13.0.0

MonoidMusician committed:
Remove hash from packages.dhall

thomashoneyman committed:
Update CI to use 0.14.0-rc5

thomashoneyman commented:
@MonoidMusician CI was failing because of the hash (as you noticed) as well as an old release candidate of the compiler. I've bumped that for you 😄

MonoidMusician committed:
Test toUpper(Full), toLower(Full)

MonoidMusician committed:
Use -Simple suffix instead of -Full

With the understanding that users should prefer the full variants.

Also add docs to the String versions and add case folding.

MonoidMusician committed:
Fix examples in docs

MonoidMusician committed:
Explicit export lists, remove FFI

thomashoneyman commented:
@MonoidMusician CI is failing because you've removed all JavaScript files under `src`. If there will no longer be FFI in this project, then you can:

- delete the `package.json` file
- delete the `.eslintrc.json` file
- replace the `.github/workflows/ci.yml` file with the standard one for libraries not using JavaScript:
  https://github.com/purescript-contrib/purescript-machines/blob/main/.github/workflows/ci.yml

That will return this library to the standard setup used in `contrib` libraries when they don't have FFI.

MonoidMusician committed:
Fix CI

MonoidMusician committed:
Update changelog

thomashoneyman committed:
Update .gitignore

thomashoneyman commented on src/Data/String/Unicode.purs:
```diff
@@ -0,0 +1,88 @@
+conv f = toCodePointArray >>> map f >>> fromCodePointArray
+
+convFull :: (CodePoint -> Array CodePoint) -> String -> String
+convFull f = toCodePointArray >>> bindFlipped f >>> fromCodePointArray
```
If `conv` stands for `convert`, can we use the full name?

```suggestion
convert :: (CodePoint -> CodePoint) -> String -> String
convert f = toCodePointArray >>> map f >>> fromCodePointArray

convertFull :: (CodePoint -> Array CodePoint) -> String -> String
convertFull f = toCodePointArray >>> bindFlipped f >>> fromCodePointArray
```

thomashoneyman commented on src/Data/String/Unicode.purs:
```diff
@@ -0,0 +1,88 @@
+-- | letter, if any.  (Title case differs from upper case only for a small
+-- | number of ligature letters.)
+-- | Any other character is returned unchanged.
+--toTitle :: CodePoint -> CodePoint
+--toTitle = conv CP.toTitle
```
Is this meant to be commented out? Should it be removed if so?

MonoidMusician replied:
Yeah, I'll just remove it. There's no clear way to divide strings into words to do proper titlecasing – if users need that they can implement an appropriate algorithm themselves using the functions in `Data.CodePoint.Unicode`.

thomashoneyman commented on src/Data/String/Unicode.purs:
```diff
@@ -0,0 +1,88 @@
+toLowerSimple :: String -> String
+toLowerSimple = conv CP.toLowerSimple
+
+-- | Code-point-to-code-point case folding. May be faster than `caseFold`.
```
Could we add a brief note describing in what situations this may be faster than `caseFold`?

thomashoneyman commented on src/Data/String/Unicode.purs:
```diff
@@ -0,0 +1,88 @@
+caseFold = convFull CP.caseFold
+
+-- | Caseless matching, based on `caseFold`.
+caselessMatch :: String -> String -> Boolean
```
Should this perhaps be `caselessEq` as this function checks equality? If `match` is the more correct term then feel free to resolve this without changes.

MonoidMusician replied:
The Unicode standards use the term caseless matching, but `caselessEq` might be more clear. I don't have a strong opinion.

thomashoneyman replied:
OK, I think it's reasonable to stay with the standard. Thanks!

thomashoneyman commented on src/Data/String/Unicode.purs:
```diff
@@ -0,0 +1,88 @@
+-- | and may map single code points to more than one code point. For example,
+-- | `toLower "\x0130" == "\x0069\x0307"`.
+-- |
+-- | Because this matches on more rules, it may be slower than `toLower`, but it
```
```suggestion
-- | Because this matches on more rules, it may be slower than `Data.CodePoint.Unicode.toLower`, but it
```

thomashoneyman commented on src/Data/String/Unicode.purs:
```diff
@@ -0,0 +1,88 @@
+-- | and may map single code points to more than one code point. For example,
+-- | `toUpper "ß" == "SS"`.
+-- |
+-- | Because this matches on more rules, it may be slower than `toUpper`, but it
```
```suggestion
-- | Because this matches on more rules, it may be slower than `Data.CodePoint.Unicode.toUpper`, but it
```

MonoidMusician replied:
Oh good catch, I actually meant `toUpperSimple` here. (Forgot to update the comments when I switched names.)

thomashoneyman commented on src/Data/String/Unicode.purs:
```diff
@@ -0,0 +1,88 @@
+modify = unsafeCoerce
+
+modifyFull :: (Int -> Array Int) -> (CodePoint -> Array CodePoint)
+modifyFull = unsafeCoerce
```
These don't seem to be used anywhere. Can they be removed?

thomashoneyman commented on unicode-version:
```diff
@@ -0,0 +1 @@
+13.0.0
```
Could we add a line to the README file that specifies we're supporting Unicode 13.0.0? We could include the link:

```md
This version of `unicode` supports the [Unicode standard, version 13.0.0](https://www.unicode.org/versions/Unicode13.0.0).
```

thomashoneyman commented on src/Data/CodePoint/Unicode.purs:
```diff
@@ -22,16 +22,20 @@ module Data.Char.Unicode
   , isNumber
 
--- Conversion to Int
+  , digitToInt
```
Can we preserve this comment?

```suggestion
  , digitToInt -- Deprecated
```

thomashoneyman commented on src/Data/CodePoint/Unicode.purs:
```diff
@@ -9,10 +9,10 @@ module Data.Char.Unicode
   , isAlphaNum
   , isLetter
-  , isHexDigit
+  , isDigit
```
Can we preserve this comment?

```suggestion
  , isDigit -- Deprecated
```

thomashoneyman commented on .github/workflows/ci.yml:
```diff
@@ -25,29 +25,11 @@ jobs:
           path: |
             .spago
             output
+      - name: Install dependencies
```
```suggestion

      - name: Install dependencies
```

thomashoneyman replied:
(Just a minor suggestion to preserve the newline)

thomashoneyman commented on src/Data/CodePoint/Unicode/Internal.purs:
```diff
@@ -1,10 +1,10 @@
 -----------------------------------------------------------
 -- This is an automatically generated file: do not edit
--- Generated by ubconfc at Fri Nov 10 20:05:16 PST 2017
+-- Generated by ubconfc at Tue Jan 12 18:57:20 CST 2021
```
I don't have a good way to verify this file other than seeing that the tests still pass. From a scan through it looks fine to me. If you have any suggestions on verifying this, @MonoidMusician, I'm all ears!

thomashoneyman replied:
Also, we have some documentation that describes generating these files anew:

https://github.com/purescript-contrib/purescript-unicode/tree/main/docs#generating-internal-modules

You've made some changes to how these are generated; if there's any information you think would be helpful to maintainers working on this in the future, please consider adding it to that documentation file as well. Thanks!

MonoidMusician replied:
Thanks, I missed that it was moved to there, will update.

thomashoneyman commented on fullcase.js:
```diff
@@ -0,0 +1,117 @@
+import Data.CodePoint.Unicode.Internal (bsearch, uTowlower, uTowtitle, uTowupper)
+import Data.Maybe (Maybe(..))
+
+type CaseRec =
```
Could we add documentation comments to this and the declarations throughout the file, as this is a public module? Simple ones are fine, but they can help guide maintainers in the future as well.

MonoidMusician replied:
I moved it to a new Internal folder – users should not use these functions on `Int`s but instead the ones on `CodePoint`s.

thomashoneyman reviewed (changes requested):
This is really impressive, @MonoidMusician, and thank you for your work. I only have minor comments to make; on the whole it's good to see this modernized and ready for Unicode 13.0.0, as well as on a sounder footing with the switch to code points.

Admittedly, I'm not particularly well-versed in the subtleties of unicode, so perhaps if @michaelficarra has time for another look that would be helpful as well.

MonoidMusician committed:
Apply suggestions from code review

Co-authored-by: Thomas Honeyman <admin@thomashoneyman.com>

MonoidMusician committed:
Move Casing to Internal module

And update documentation for generating internal modules

MonoidMusician committed:
conv -> convert

Co-authored-by: Thomas Honeyman <admin@thomashoneyman.com>

MonoidMusician committed:
Address more review suggestions

MonoidMusician committed:
Update README

MonoidMusician committed:
Mention that *Simple variants preserve the number of code points

Instead of hypothetical performance benefits

MonoidMusician committed:
Merge remote-tracking branch 'purescript-contrib/main'

MonoidMusician committed:
Oops, need to rename usages of conv -> convert

MonoidMusician commented:
Okay, thanks for the feedback! I think I addressed what was addressable.

thomashoneyman commented on packages.dhall:
```diff
@@ -1,4 +1,4 @@
 let upstream =
-      https://raw.githubusercontent.com/purescript/package-sets/prepare-0.14/src/packages.dhall
+      https://raw.githubusercontent.com/purescript/package-sets/prepare-0.14/src/packages.dhall sha256:baefc2d9387175879c9576620d245e708b11cda41569aac9631b351cb0616d06
```
```suggestion
      https://raw.githubusercontent.com/purescript/package-sets/prepare-0.14/src/packages.dhall
```

thomashoneyman committed:
Update packages.dhall

thomashoneyman reviewed (approved):
👍 Looks good to me! Thanks for your work! If no other maintainer notices an issue before this must be merged in order to make the 0.14 release, then I will merge this PR. I'm going to leave it open in the meantime.

michaelficarra commented on fullcase.js:
```diff
@@ -0,0 +1,117 @@
+const scData = {};
+const cfData = {};
+
+const rhex = a => typeof a === 'string' ? +('0x'+a) : +a;
```
```suggestion
const rhex = a => parseInt(a, 16);
```

Do we expect any of these keys to not be strings?

michaelficarra replied:
```suggestion
const rhex = a => parseInt(a, 16);
```

Do we expect any of these keys to not be strings?

michaelficarra commented on fullcase.js:
```diff
@@ -0,0 +1,117 @@
+const rhex = a => typeof a === 'string' ? +('0x'+a) : +a;
+const sortHex = (a,b) => rhex(a) - rhex(b);
+
+for (const spec of SpecialCasing.matchAll(special)) {
```
`String.prototype.matchAll` is a very recent feature. What kind of engine support are we targeting with this project? We can accomplish the same thing without it if we need.

**edit:** Since this is a development time script, I guess we can assume a very modern engine.

MonoidMusician replied:
This is just for generating code, it's not part of the runtime of the library. So I assume that developers can find a recent enough node.

michaelficarra commented on README.md:
```diff
@@ -5,7 +5,11 @@
+
+General functions for working with `String`s in terms of `CodePoint`s are found in the `Data.String.CodePoints` module in the [`strings` library](https://github.com/purescript/purescript-strings).
+
+This version of the `unicode` library supports the [Unicode standard, version 13.0.0](https://www.unicode.org/versions/Unicode13.0.0).
```
I would just link to `unicode-version` and say that the currently supported version can be found in that file. We plan on updating this project following the release of a new Unicode version, right?

michaelficarra commented on fullcase.js:
```diff
@@ -0,0 +1,117 @@
+  `{ code: ${d.code}, lower: ${d.lower}, title: ${d.title}, upper: ${d.upper}, fold: ${d.fold}, foldFull: ${d.foldFull} }`
+);
+
+const file = `module Data.CodePoint.Unicode.Internal.Casing where
```
The first line of generated files should be a comment that says it was generated and what script generated it. Additionally, you should mark generated files in the `.gitattributes` so that GitHub suppresses their diffs. See https://github.com/github/linguist#generated-code

michaelficarra reviewed (commented):
Generally LGTM. What do you think about also exporting the Unicode version as a String?

MonoidMusician committed:
Use parseInt

Co-authored-by: Michael Ficarra <github@michael.ficarra.me>

MonoidMusician committed:
Add linguist-generated attribute

MonoidMusician committed:
Add header to generated file

MonoidMusician committed:
Reference unicode-version

MonoidMusician commented:
Okay, I think I addressed your feedback, thanks! I don't think we need to export the Unicode version from the library, I just want to make sure it's available for documentation when someone goes looking for it. Unicode seems to be pretty stable so I don't think it affects most users.

MonoidMusician commented:
I just noticed that this fixes #28 (I used hexadecimal escapes).

thomashoneyman commented:
Fantastic work, @MonoidMusician! Thank you, and if anyone does happen to notice an issue we can merge a fix.
